File: robot_simulator/c#/robot_simulator.cs
using System;

namespace robot_simulator
{
    class MainClass
    {
        public static void Main(string[] args)
        {
            Robot r = new Robot(7, 3, "north", "RAALAL");
            r.run();
        }
    }

    enum CardinalDirections {north, east, south, west}

    class Robot
    {
        private int x { get; set; }
        private int y { get; set; }
        private CardinalDirections bearing { get; set; }
        private string instructions { get; set; }

        public Robot(int x0, int y0, string bearing, string instructions)
        {
            x = x0;
            y = y0;
            this.instructions = instructions;

            if (bearing.ToLower().Equals("north"))
                this.bearing = CardinalDirections.north;
            else if (bearing.ToLower().Equals("east"))
                this.bearing = CardinalDirections.east;
            else if (bearing.ToLower().Equals("south"))
                this.bearing = CardinalDirections.south;
            else
                this.bearing = CardinalDirections.west;
        }

        private void advance()
        {
            switch (bearing)
            {
                case CardinalDirections.north:
                    y++;
                    break;
                case CardinalDirections.east:
                    x++;
                    break;
                case CardinalDirections.south:
                    y--;
                    break;
                case CardinalDirections.west:
                    x--;
                    break;
            }
        }

        private void turn(char dir)
        {
            int currentBearing = (int)bearing;
            if (dir == 'R' || dir == 'r')
                bearing = (CardinalDirections)((currentBearing + 1) % 4);
            else
                bearing = (CardinalDirections)((currentBearing - 1) % 4);

            if ((int)bearing < 0)
                bearing = (CardinalDirections)((int)bearing + 4);
        }

        public void run()
        {
            foreach (char c in instructions)
            {
                if (c == 'A' || c == 'a')
                    advance();
                else
                    turn(c);
            }
            Console.WriteLine("Final Position: (" + x + "," + y + ")");
            Console.WriteLine("Final Bearing: " + bearing.ToString());
        }
    }
}

File: robot_simulator/java/RobotSimulator.java
/**
 * 
 * @author Sagar Rathod
 * @version 1.0
 */
public class RobotSimulator {

	enum Direction {

		NORTH, EAST, SOUTH, WEST;

		Direction left, right;

		public void setLeft(Direction left) {
			this.left = left;
		}

		public void setRight(Direction right) {
			this.right = right;
		}

		public Direction getLeft() {
			return left;
		}

		public Direction getRight() {
			return right;
		}
	}

	final static Direction NORTH = Direction.NORTH;
	final static Direction EAST = Direction.EAST;
	final static Direction SOUTH = Direction.SOUTH;
	final static Direction WEST = Direction.WEST;

	static void initSimulator() {
		
		NORTH.setLeft(Direction.WEST);
		NORTH.setRight(Direction.EAST);

		EAST.setLeft(Direction.NORTH);
		EAST.setRight(Direction.SOUTH);

		SOUTH.setLeft(Direction.EAST);
		SOUTH.setRight(Direction.WEST);

		WEST.setLeft(Direction.SOUTH);
		WEST.setRight(Direction.NORTH);

	}

	static class Robot {

		private int x;
		private int y;
		private Direction faceDirection;

		public Robot(int x, int y, Direction faceDirection) {
			
			this.x = x;
			this.y = y;
			this.faceDirection = faceDirection;
		}

		public void advance() {

			if (faceDirection == Direction.EAST)
				x++;

			if (faceDirection == Direction.WEST)
				x--;

			if (faceDirection == Direction.NORTH)
				y++;

			if (faceDirection == Direction.SOUTH)
				y--;
		}

		public void turnLeft() {
			faceDirection = faceDirection.getLeft();
		}

		public void turnRight() {
			faceDirection = faceDirection.getRight();
		}

		@Override
		public String toString() {
			return "Robot is standing at (x, y) =" + " (" + x + ", " + y + ") position facing " + faceDirection
					+ " direction.";
		}

		/**
		 * Simulates robot walking path.
		 * 
		 * @param path
		 */
		public void simulateRobotWalkingPath(String path) {

			System.out.println("Initial Position:" + this);
			char[] chars = path.toCharArray();

			for (int i = 0; i < chars.length; i++) {
				if (chars[i] == 'R') {
					turnRight();
				}

				if (chars[i] == 'L') {
					turnLeft();
				}

				if (chars[i] == 'A') {
					advance();
				}
			}
			System.out.println("Final Position:" + this);
		}
	}

	public static void main(String[] args) {

		initSimulator();
		Robot robot = new Robot(7, 3, NORTH);
		robot.simulateRobotWalkingPath("RAALAL");

		robot = new Robot(4, 5, SOUTH);
		robot.simulateRobotWalkingPath("LAARALL");
	}

}
